src/librustc_resolve/lib.rs

```diff
@@ -395,12 +395,19 @@ enum AliasPossibility {
     Maybe,
 }
 
+/// Is the `PathSource::Trait` from an `impl` or a bound?
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+enum ImplOrBound {
+    Impl,
+    Bound,
+}
+
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
 enum PathSource<'a> {
     // Type paths `Path`.
     Type,
     // Trait paths in bounds or impls.
-    Trait(AliasPossibility),
+    Trait(AliasPossibility, ImplOrBound),
```

Comment:
Wait.
`AliasPossibility` is supposed to be exactly the same thing as `ImplOrBound` - aliases can be used in bounds, but cannot be used in `impl Trait for ...`.
Also, poly-trait-ref is *always* a bound, no need to change visitors.
What this PR tries to achieve?

Reply:
Thanks @petrochenkov ! That's really useful info. The goal was simply to emit a different error if something that is not a trait is used in a bound as opposed to an impl (i.e. `impl Struct for Foo` vs `foo<T: Struct>(...) {}`).

Reply:
So you are saying that if `AliasPossibility::Maybe` it _must_ be a bound. But is it also true that if `AliasPossibility::No` it _must_ be an `impl`?

Reply:
@petrochenkov Is this correct:

```diff
diff --git a/src/libsyntax/visit.rs b/src/libsyntax/visit.rs
index bf8beb1..5492cd6 100644
--- a/src/libsyntax/visit.rs
+++ b/src/libsyntax/visit.rs
@@ -548,9 +541,9 @@ impl<'a> PathSource<'a> {
         __diagnostic_used!(E0578);
         __diagnostic_used!(E0909);
         match (self, has_unexpected_resolution) {
-            (PathSource::Trait(_, ImplOrBound::Impl), true) => "E0404",
-            (PathSource::Trait(_, ImplOrBound::Bound), true) => "E0909",
-            (PathSource::Trait(_, _), false) => "E0405",
+            (PathSource::Trait(AliasPossibility::No), true) => "E0404",
+            (PathSource::Trait(AliasPossibility::Maybe), true) => "E0909",
+            (PathSource::Trait(_), false) => "E0405",
             (PathSource::Type, true) => "E0573",
             (PathSource::Type, false) => "E0412",
             (PathSource::Struct, true) => "E0574",
```

Reply:
No, the existing code is already correct, only `E0245` need to be removed as dead code.
Also docs for `E0404` should probably be tweaked to mention bounds in addition to impls.

Reply:
To clarify, https://github.com/rust-lang/rust/pull/48446#discussion_r170138387 is correct too, if you really want to make these error codes more fine grained, I'm just not sure this makes sense given that trait aliases are not even implemented yet.

If you do want to spend time on this, I'd suggest to split `E0405` as well and document differences between bounds and impls and how trait aliases can be used in bounds but not in impls in the descriptions of `E0404`/`E0405`/`E0909`/`E0910`.

Reply:
>  if you really want to make these error codes more fine grained

I think that was the conclusion of #36337

> If you do want to spend time on this, I'd suggest to split E0405 as well and document differences between bounds and impls and how trait aliases can be used in bounds but not in impls in the descriptions of E0404/E0405/E0909/E0910.

Sure, I can try to do that.